File: 297-serialize-and-deserialize-binary-tree/serialize-and-deserialize-binary-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

/*
 * Encodes a tree to a single string.
 */
function serialize(root: TreeNode | null): string {
    const pre = []

    const dfs = (root) => {
        if (root === null) {
            pre.push('#')
            return
        }

        pre.push(root.val)
        dfs(root.left)
        dfs(root.right)
    }
    dfs(root)

    return pre.join(",")
};

/*
 * Decodes your encoded data to tree.
 */
function deserialize(data: string): TreeNode | null {
    const pre = data.split(",")
    let i = 0
    const dfs = () => {
        if (pre[i] == '#') {
            i++
            return null
        }
        const node = new TreeNode(Number(pre[i]))
        i++
        node.left = dfs();
        node.right = dfs();
        return node
    }
    
    return dfs()
};


/**
 * Your functions will be called as such:
 * deserialize(serialize(root));
 */

File: 297-serialize-and-deserialize-binary-tree/serialize-and-deserialize-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        # make pre and inorder
        pre = []

        def dfs(root):
            if root is None:
                pre.append("#")
                return
            
            pre.append(str(root.val))
            dfs(root.left)
            dfs(root.right)
        dfs(root)
        return ",".join(pre)

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        nodes = data.split(',')
        self.i = 0

        def dfs():
            if nodes[self.i] == "#":
                self.i +=1
                return None
            
            root = TreeNode(nodes[self.i])
            self.i += 1
            root.left = dfs()
            root.right = dfs()
            return root
        return dfs()


# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))